ReactQt/tests/test-netexecutor-socket.cpp

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright (c) 2017-present, Status Research and Development GmbH.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <QScopedPointer>
+#include <QSignalSpy>
+#include <QTest>
+#include <QVariant>
+#include <QtMath>
+#include <QtQuick/QQuickView>
+
+#include "bridge.h"
+#include "netexecutor.h"
+#include "reacttestcase.h"
+#include <QTcpServer>
+
+const int TEST_EXECUTOR_PORT = 3000;
+const int MAX_PACKAGES_COUNT = 9;
+const int CHUNK_SIZE = 100000;
+
+class TestNetExecutorSocket : public ReactTestCase {
+    Q_OBJECT
+
+protected:
+    void sendPackageInTwoWrites(quint32 bytesCount);
+    void sendPackageInOneWrite(quint32 bytesCount);
+    void sendPackageInChunks(quint32 bytesCount);
+    void sendTwoPackagesInOneWrite(quint32 package1bytesCount, quint32 package2bytesCount);
+    QByteArray packageData(int bytesCount);
+    QByteArray packageHeader(quint32 bytesCount);
+    QByteArray package(quint32 bytesCount);
+    quint32 packageSize(quint32 packageNumber);
+    void waitAndCheckPackagesReceived(QSignalSpy& spy, int count);
+
+private slots:
+    void initTestCase() override;
+
+    void testSendingPackagesInTwoWrites();
+    void testSendingPackagesInOneWrite();
+    void testSendingPackagesInChunks();
+    void testSendingTwoPackagesAtOnce();
+
+private:
+    QSharedPointer<NetExecutor> m_netExecutor;
+    QSharedPointer<QTcpServer> m_server;
+    QSharedPointer<QTcpSocket> m_socket;
+};
+
+void TestNetExecutorSocket::initTestCase() {
+
+    m_server = QSharedPointer<QTcpServer>(new QTcpServer());
+    m_server->listen(QHostAddress::LocalHost, TEST_EXECUTOR_PORT);
+
+    m_netExecutor = QSharedPointer<NetExecutor>(new NetExecutor);
+    m_netExecutor->setPort(TEST_EXECUTOR_PORT);
+    m_netExecutor->init();
+
+    waitAndVerifyCondition([=]() { return m_server->hasPendingConnections(); },
+                           "NetExecutor didn't connect to Tcp server");
+    m_socket = QSharedPointer<QTcpSocket>(m_server->nextPendingConnection());
+    QVERIFY(m_socket->isValid());
+}
+
+void TestNetExecutorSocket::sendPackageInTwoWrites(quint32 bytesCount) {
+    m_socket->write(packageHeader(bytesCount));
+    m_socket->write(packageData(bytesCount));
+    QCoreApplication::processEvents();
+}
+
+void TestNetExecutorSocket::sendPackageInOneWrite(quint32 bytesCount) {
+    m_socket->write(package(bytesCount));
+    QCoreApplication::processEvents();
+}
+
+void TestNetExecutorSocket::sendPackageInChunks(quint32 bytesCount) {
+    QByteArray ba = package(bytesCount);
+
+    while (ba.length()) {
+        QByteArray chunk = ba.left(CHUNK_SIZE);
+        ba.remove(0, chunk.length());
+        m_socket->write(chunk);
```

Comment:
`m_socket` probably should be flushed each time when you expect that data should be really sent.
To simulate situation to put multiple packets into socket buffer, maybe, make sense to extend methods "sendPackageXXX" with additional input argument `bool flushSocket`

Reply:
@MaxRis, explicit call to `QApplication::processEvents()` has the same effect as socket flushing, since data sent to network whenever app process event loop. Probably it would be more clear to call `flush()` explicitly instead, thanks.
I think that additional argument won't be helpful since I want socket to be flushed every time I sent test data.